### PR TITLE
feat(launch): enable launch macros for all runners

### PR DESCRIPTION
**Pull request by bcsherma**
Fixes
-----

Fixes WB-13828

Fixes #NNNN

Description
-----------
What does the PR do?

### Summary
🐳🚀🔧

Add macro feature to launch runners. This feature allows users to customize the docker image URI for different launch runners by using placeholders in the `docker_image` attribute of the launch configuration files. The changes apply the `fill_macros` method to the `docker_image` or `image_uri` values for the Kubernetes, local container, SageMaker, and Vertex runners.

> _`fill_macros` method_
> _customizes image URI_
> _for each runner type_

### Walkthrough
*  Add `fill_macros` method to `launch_project` class to replace placeholders in strings with values from the object ([link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86R456), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86R468), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83R114), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83R136), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7R156), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7R166), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-971309ad05d166697209cdb39e7af4378445d88f459aadd134a5509f7fd70a62L140-R140))
*  Use `fill_macros` method to construct and validate image URIs for different runners from `docker_image` attribute and `launch_project` name in `kubernetes_runner.py`, `local_container.py`, `sagemaker_runner.py`, and `vertex_runner.py` ([link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86R456), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-8ea6631c99cbc07d88e28024cec144305c704436041cc90266e8ff10c2017f86R468), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83R114), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-03853de6b6ac99be1dc94a98df47da16d8d172f5870a0fbbe5aa39e59e89df83R136), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7R156), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-66f6587e310db906417a8cc4b6f996469d58ab461405436c713217d850726ca7R166), [link](https://github.com/wandb/wandb/pull/5624/files?diff=unified&w=0#diff-971309ad05d166697209cdb39e7af4378445d88f459aadd134a5509f7fd70a62L140-R140))

Testing
-------
How was this PR tested?

Checklist
-------
- [ ] Include reference to internal ticket "Fixes WB-NNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 638974e</samp>

> _`fill_macros` method_
> _customizes image URI_
> _for each runner type_

### CODE REVIEW

File: wandb/sdk/launch/runner/sagemaker_runner.py [hunk 1]
```diff
@@ -162,6 +163,7 @@ def run(
                 launch_project,
                 entry_point,
             )
+            launch_project.fill_macros(image)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/sagemaker_runner.py:166]: Why not jus do this once outside the if statement?

**Review by KyleGoyette — APPROVED** [review]
nit: In a lot of these cases you could just do this once no?

File: wandb/sdk/launch/runner/sagemaker_runner.py (hunk 1 above)
Comment (bcsherma) [reply wandb/sdk/launch/runner/sagemaker_runner.py:166]: u r smart!

File: wandb/sdk/launch/runner/local_container.py [hunk 2]
```diff
@@ -132,6 +133,7 @@ def run(
                 launch_project,
                 entry_point,
             )
+            launch_project.fill_macros(image_uri)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/local_container.py:136]: Seems like we could do it in one case here too

File: wandb/sdk/launch/runner/kubernetes_runner.py [hunk 3]
```diff
@@ -471,6 +472,7 @@ def _inject_defaults(
             assert entry_point is not None
             assert builder is not None
             image_uri = builder.build_image(launch_project, entry_point)
+            launch_project.fill_macros(image_uri)
```
Comment (KyleGoyette) [inline wandb/sdk/launch/runner/kubernetes_runner.py:475]: and in one case here?

**Pull request merged**